### PR TITLE
Ensure task_instance exist before running update on its state(REST API)

### PR DESCRIPTION
**Pull request by ephraimbuddy**
Error when the task instance does not exist

Related: https://github.com/apache/airflow/issues/18579
---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/api_connexion/endpoints/task_instance_endpoint.py [hunk 1]
```diff
@@ -286,9 +286,14 @@ def post_set_task_instances_state(dag_id, session):
         error_message = f"Task ID {task_id} not found"
         raise NotFound(error_message)
 
+    execution_date = data['execution_date']
+    ti = session.query(TI).filter_by(execution_date=execution_date, task_id=task_id, dag_id=dag_id).first()
```
Comment (ashb) [inline airflow/api_connexion/endpoints/task_instance_endpoint.py]:
```suggestion
    ti = session.query(TI).filter_by(execution_date=execution_date, task_id=task_id, dag_id=dag_id).one_or_none()
```

**Review by ashb — COMMENTED** [review]
In a new PR we should add `run_id` to this API endpoint too.

File: tests/api_connexion/endpoints/test_task_instance_endpoint.py [hunk 2]
```diff
@@ -1129,6 +1129,37 @@ def test_should_raise_404_not_found_dag(self):
...
+    @mock.patch('airflow.models.dag.DAG.set_task_instance_state')
+    def test_should_raise_not_found_if_execution_date_is_wrong(self, mock_set_task_instance_state, session):
+        self.create_task_instances(session)
+        mock_set_task_instance_state.return_value = (
+            session.query(TaskInstance)
+            .join(TaskInstance.dag_run)
+            .options(contains_eager(TaskInstance.dag_run))
+            .filter(TaskInstance.task_id == "print_the_context")
+            .all()
+        )
```
Comment (jedcunningham) [inline tests/api_connexion/endpoints/test_task_instance_endpoint.py]: Do we need to set up the mock return since we are short circuiting anyways?

Comment (ephraimbuddy) [reply tests/api_connexion/endpoints/test_task_instance_endpoint.py]: We don't need actually, even the mock itself. I will remove the return value

Comment (jedcunningham) [reply tests/api_connexion/endpoints/test_task_instance_endpoint.py]: Keeping the mock itself may make sense so you can ensure it wasn't actually called 👍

**Review by jedcunningham — APPROVED** [review]

File: airflow/api_connexion/endpoints/task_instance_endpoint.py [hunk 3]
```diff
@@ -286,9 +286,14 @@ def post_set_task_instances_state(dag_id, session):
         error_message = f"Task ID {task_id} not found"
         raise NotFound(error_message)
 
+    execution_date = data['execution_date']
+    ti = session.query(TI).filter_by(execution_date=execution_date, task_id=task_id, dag_id=dag_id).one_or_none()
+    if not ti:
+        raise NotFound(f"Task instance not found for task {task_id} on execution_date {execution_date}")
```
Comment (uranusjr) [inline airflow/api_connexion/endpoints/task_instance_endpoint.py]: Probably can just be

```python
try:
    ti = session.query(TI)...one()
except NoResultFound:
    raise NotFound(...)

**Review by uranusjr — APPROVED** [review]
Almost commented about adding `run_id` as well…

**Comment by ephraimbuddy** [comment]
> In a new PR we should add `run_id` to this API endpoint too.

Added: https://github.com/apache/airflow/pull/18724

**Review by kaxil — APPROVED** [review]

**Pull request merged**